plugins/fetcher/prometheus/metadata.go

```diff
@@ -0,0 +1,69 @@
+// Licensed to Apache Software Foundation (ASF) under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Apache Software Foundation (ASF) licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package prometheus
+
+import (
+	"errors"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/scrape"
+)
+
+// MetadataCache is an adapter to prometheus' scrape.Target  and provide only the functionality which is needed
+type MetadataCache interface {
+	Metadata(metricName string) (scrape.MetricMetadata, bool)
+	SharedLabels() labels.Labels
+}
+
+type ScrapeManager interface {
+	TargetsAll() map[string][]*scrape.Target
+}
+
+type metadataService struct {
+	sm ScrapeManager
+}
+
+func (s *metadataService) Get(job, instance string) (MetadataCache, error) {
+	targetGroup, ok := s.sm.TargetsAll()[job]
+	if !ok {
+		return nil, errors.New("unable to find a target group with job=" + job)
+	}
+
+	// from the same targetGroup, instance is not going to be duplicated
+	for _, target := range targetGroup {
+		if target.Labels().Get(model.InstanceLabel) == instance {
+			return &mCache{target}, nil
+		}
+	}
+
+	return nil, errors.New("unable to find a target with job=" + job + ", and instance=" + instance)
+}
+
+// mCache as an adapter to get metadata from scrape.Target
+type mCache struct {
+	t *scrape.Target
+}
+
+func (m *mCache) Metadata(metricName string) (scrape.MetricMetadata, bool) {
+	return m.t.Metadata(metricName)
```

Comment:
may cause panic

Reply:
The `t` is set before (https://github.com/apache/skywalking-satellite/pull/42/files/0f01d525e7dc208fff2573ae171c330f84481175#diff-ca991bdd94c1589afcfcfdbe1bb4068b2ca7b3421363e8b53e92bfc477e66df8R51)